site/bps/BP-22-separate-closing-ledgers-from-opening-ledgers.md

```diff
@@ -0,0 +1,97 @@
+---
+title: "BP-22: Separate closing ledgers from opening ledgers"
+issue: https://github.com/apache/bookkeeper/794
+state: "Under Discussion"
+release: "4.7.0"
+---
+
+### Motivation
+
+In the beginning there was openLedger. Opening a ledger meant also figuring out what the last entry of a ledger should be, and writing it to ZooKeeper. For a long time this was the only way to read a ledger. If a writer was writing to the ledger, then anything it wrote after this point would be lost (fencing was added later). The open operation was the natural place to put recovery, as at this point it was only possible to read a recovered ledger.
+
+openLedgerNoRecovery was added in 2011. This allowed users to read from a ledger as it was being written to, which opened up a bunch of tailing uses cases. Recovery was still the default, because that was what it had always been, and tailing was still considered a secondary usecase. If the user wanted to skip recovery, they'd have to explicitly call the no recovery method.
+
+Then the new API arrived. In the new API, tailing is the primary read use case, and recovery has been demoted to a boolean flag on a builder for the open operation. The user is for the most part unaware of recovery.
+
+However, recovery is still one of the most important aspects of BookKeeper. It is the mechanism on which our Total Order Atomic Broadcast guarantees are built. It deserves to be a bit more prominent than a boolean flag in a builder. It also doesn't help that the terminology is inconsistent. The flag is called withRecovery, while the to check if recovery is needed, we call isClosed. Closed itself is ambiguous because it may refer to the local handle, or it may refer to the state of the ledger.
+
+As tailing is now the primary usecase, we expect that if a writer fails, then whichever node takes over as writer already has a non-recovered ReadHandle open. It would be nice to be able to continue using this Handle to read to the end of the ledger.
+
+### Public Interfaces
+
+I proposes the removal of OpenOpBuilder#withRecovery()
+
+A new method on BookKeeper:
+```
+interface BookKeeper {
+
+    ...
+
+    CompletableFuture<Void> seal(ReadHandle handle);
```

Comment:
why not
`CompletableFuture<Void> seal(long ledgerId)`
or
`CompletableFuture<Void> newSealOp().withLedgerId(ledgerId).execute();`

Maybe we we could return the LAC instead of Void.

The user need to re-open the ledger but it is clearer

Reply:
The user needs more than the ledgerId to seal the ledger. They need digest type and password at least. In the case of log0 they'll also need the ensemble. It's more foolproof to pull this from a ReadHandle (via getLedgerMetadata) than to require the user to pass each one in.

There's no need to reopen. It should be possible to reread the LAC and updated metadata with the existing handle.

Reply:
ok.
As WriteHandle extends ReadHandle this new API can be used with a an existing WriteHandle, we should forbid this.

counter proposal:
1) drop OpenBuilder#withRecovery(true)
2) add OpenBuilder#sealing()

Reply:
> As WriteHandle extends ReadHandle this new API can be used with a an existing WriteHandle, we should forbid this.

Why? We'd be performing a seal operation on a Ledger in either case.

> counter proposal:
>    drop OpenBuilder#withRecovery(true)
>    add OpenBuilder#sealing()

I had considered something like this. It's better than withRecovery(), but you still need to reopen the ledger, sealing is still hidden inside a builder, and it makes open operation with possible side effects, which I'd like to avoid (similar to not putting #forceSealed on ReadHandle).

Reply:
I think what Enrico proposed with a separate `seal` call makes more sense to me. In some cases (e.g. admin cases), you can use this seal call to seal a ledger.

changing `withRecovery(true)` to `#sealing()` works for me as well. but I would prefer not removing `withRecovery(true)` or `#sealing()` from the open builder, we should have this flag in the builder for opening ledger to seal/fence.

Reply:
> I think what Enrico proposed with a separate seal call makes more sense to me. In some cases (e.g. admin cases), you can use this seal call to seal a ledger.
What do you mean by a separate seal call?

> changing withRecovery(true) to #sealing() works for me as well. but I would prefer not removing withRecovery(true) or #sealing() from the open builder, we should have this flag in the builder for opening ledger to seal/fence.

Why?

In any case, if we do leave it in the builder, we should call it #forceSealed or #forceClosed, to inform the user that the option has heavy side effects. In fact, I think any method on BookKeeper should be called the same.

Reply:
@ivankelly 

I mean what Enrico proposed `CompletableFuture<Void> seal(long ledgerId)` is making more sense to me than `CompletableFuture<Void> seal(ReadHandle rh)`. `seal` call is separated from `open` from this point of view.

I am fine with calling it `forceSealed` or `forceClosed` in the open builder. as I said in the other comment: "I think open-to-seal is more naturally fit in open builder, as a boolean flag (withRecovery(true)), or as a OpenFlag (O_SEAL or O_FENCE or O_RECOVERY or whatever), or whatever name making sense (#sealing() as what @eolivelli proposed)".